File: cpp/Problem75.cpp
/*
 * Problem: 75. Sort Colors
 * URL: https://leetcode.com/problems/sort-colors/
 * Difficulty: Medium
 * Tags: Array, Dutch Flag Algorithm
 */

#include <vector>

using namespace std;

class Solution {
public:
	void sortColors(vector<int>& nums) {
		int left = 0, mid = 0, right = nums.size() - 1;
		while (mid <= right) {
			if (nums[mid] < 1) {
				swap(nums[left], nums[mid]);
				left++;
				mid++;
			} else if (nums[mid] > 1) {
				swap(nums[mid], nums[right]);
				right--;
			} else {
				mid++;
			}
		}
	}

	void swap(int &a, int &b) {
		int tmp = a;
		a = b;
		b = tmp;
	}
};

File: python/Problem75.py
# Problem: 75. Sort Colors
# URL: https://leetcode.com/problems/sort-colors/
# Difficulty: Medium
# Tags: Array, Dutch Flag Algorithm

from typing import List

class Solution:
  def sortColors(self, nums: List[int]) -> None:
    left: int = 0
    mid: int = 0
    right: int = len(nums) - 1
    while mid <= right:
      if nums[mid] < 1:
        nums[left], nums[mid] = nums[mid], nums[left]
        left += 1
        mid += 1
      elif nums[mid] > 1:
        nums[mid], nums[right] = nums[right], nums[mid]
        right -= 1
      else:
        mid += 1